cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -166,32 +168,48 @@ struct BooleanAnyImpl : public ScalarAggregator {
   Status MergeFrom(KernelContext*, KernelState&& src) override {
     const auto& other = checked_cast<const BooleanAnyImpl&>(src);
     this->any |= other.any;
+    this->has_nulls |= other.has_nulls;
     return Status::OK();
   }
 
-  Status Finalize(KernelContext*, Datum* out) override {
-    out->value = std::make_shared<BooleanScalar>(this->any);
+  Status Finalize(KernelContext* ctx, Datum* out) override {
+    if (!options.skip_nulls && !this->any && this->has_nulls) {
```

Comment:
Indeed this is to to make the PR "kleen" to match [R behavior](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/any). Meanwhile [Pandas' any is non-kleen](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.any.html).

```
>>> import pandas as pd
>>> pd.Series([None, None]).any(skipna=True)
False
>>> pd.Series([None, None]).any(skipna=False)
>>> 
```

We have three options IMO:
- Revert to non-kleene c++ behaviour and add a small fix to R
- Add kleen any/all kernels and route in R depending on flags
- Keep c++ as is and add a fix to Python (that could introduce a lot of new unvanted logic)

Reply:
> Meanwhile Pandas' any is non-kleen.

The pandas `any`/`all` methods were broken for object dtype (and since numpy doesn't support nulls in its boolean dtype, whenever you have missing values, you have object dtype), so best not to use that as a reference (see eg https://github.com/pandas-dev/pandas/issues/27709)

Reply:
So how about then just making this kernel "kleene" and just document that fact?

Reply:
Shall I:
1. Document the behavior as kleen?
2. Rather revert the change and first implement `al_kleen/any_kleen` ([ARROW-10291](https://issues.apache.org/jira/browse/ARROW-10291)) and then map R to those for `skip_nulls==False`?

Reply:
I am not sure there is a good use case for a non-kleene version, so I am fine with just documenting for now that the behaviour follows Kleene logic (so is the reducing version of and/or_kleene)

Reply:
I documented the new behavior in [compute.rst](https://github.com/apache/arrow/pull/10476/files#diff-ce5b94577014735990903d3d03bd4ea4b8c8e6d32f5227592e60b7dd6a912d59) and [api_aggregate.h](https://github.com/apache/arrow/pull/10476/files#diff-694a635f185cf0d22dd6eefd91acedb48ed0bf71b15fd591aa6e7c508ba4cb09).